http-clients/src/main/java/com/palantir/remoting3/clients/ClientConfigurations.java

```diff
@@ -42,6 +42,8 @@
     private static final Duration DEFAULT_WRITE_TIMEOUT = Duration.ofMinutes(10);
     private static final Duration DEFAULT_BACKOFF_SLOT_SIZE = Duration.ofMillis(500);
     private static final boolean DEFAULT_ENABLE_GCM_CIPHERS = false;
+    private static final int DEFAULT_MAX_NUM_RETRIES_LOWER_LIMIT = 3;
```

Comment:
Hmmm, not sure this is a wise limit given #610. Also, not sure why you'd want to retry 3 times if there are just 2 URLs, especially given we apply exponential backoff with a pretty low retry delay. If you upped the retry delay, then you're going to get some really bad p99 delays.

Reply:
Basically, think trying a request each time against each URI is the best you can do - if all URIs are down, then you're out of luck, you better fail the request.

Reply:
A URI can often be "down" when let's say the service is in a stale state due to which it is unable to serve the request and retuns a 503. In such a case, even if there are just 1 or 2 URLs, you'd want the request to retry as it's expected that the stale state is very brief.

Reply:
yep, it's fine to retry more than #urls times. but I don't think we should impose a lower bound.

Reply:
Wouldn't people still be able to overwrite this though?
We are strictly talking about good defaults here. Doesn't a minimum greater than one seem more appropriate?

Reply:
oh, I didn't read carefully enough, sorry.

Reply:
It's not clear that a retry of > 1 but less than the number of URIs actually is more appropriate. In particular, I think without a global timeout then we will end up being less resilient due to retry explosions deeper in a service call tree. [This link](https://developers.redhat.com/blog/2017/06/01/microservices-patterns-with-envoy-proxy-part-ii-timeouts-and-retries/) describes this and how envoy solves it.

Reply:
Going to talk with @uschi2000 offline an report back here!

Reply:
Correction: It's not clear that a retry of > 1 AND greater than the number of URIs is more appropriate**

Reply:
@uschi2000 @jared2501 any conclusions from your talks?

Reply:
No conclusion here?

Reply:
Not yet!

Reply:
This is just a change of defaults - would be surprised if this is blocking anything (you can always change your configuration defaults if you so desire).

Reply:
This isn't blocking sure, but changing the defaults in the product's shipped configuration isn't really a desirable solution. There's multiple issues there:
- If we put a default value in the shipped config, it'll no longer grow with the number of available uris
- If we do end up changing the behavior here, that would mean removing `max-num-retries` from the product's shipped config. Now at that point if people have already overridden the product default (maybe if they have more URIs), their override will become unnecessary. 

The alternative would be to encode the logic that I have in this PR in my code, but the way the service configuration is used along with internal libraries makes that code quite fragile.

Regarding the retry explosion issue that you brought up, doesn't it already exist for cases when number of URIs > 1? I'm not sure why the problem would happen only if retry > greater than number of URIs.

Reply:
@a-k-g - 

The main thing is that if retries > URLs then you're going to retry URLs that you know have failed. Currently, retrying URLs that you know have failed currently won't significantly increase the chance of success after they've been tried already since the default backoff is quite small and the number of URLs that are tried is also quite small. I.e. only milliseconds will have passed between retries.

If we were to increase the period that we wait before retring already tried URLs, then we start to have issues with threadpools building up and clogging up other requests that are coming in that won't necessarily fail. Other RPC systems appear to solve this problem by some combination of: a) a global deadline for an RPC chain, b) circuit breaking c) making the RPC calls non-blocking, so that the queues for outgoing RPCs can be much larger.

Without some combination of the above, I doubt changing the default behaviour to try a URL more than once will offer any upside, and in fact may actually be harmful.

However, I'd like to chat with @uschi2000 about this.